File: Problem09/c/PalindromeNumber.c
#include <stdio.h>
#include <stdbool.h>

bool IsPalindrome(int x)
{
    if (x < 0)
        return false;

    if (x == 0)
        return true;

    if (x % 10 == 0)
        return false;

    int temp = x;
    int reversed = 0;

    while (temp > 0)
    {
        reversed = reversed * 10 + temp % 10;
        temp /= 10;
    }

    return x == reversed;
}

int main()
{
    int x = 121;
    bool result = IsPalindrome(x);

    printf("Is %d a palindrome number? %s\n", x, result ? "Yes" : "No");
    return 0;
}

File: Problem09/PalindromeNumber.go
package main

import "fmt"

func main() {
	x := 121
	result := IsPalindrome(x)
	fmt.Println(result)
}

func IsPalindrome(x int) bool {
	if x < 0 {
		return false
	}

	if x == 0 {
		return true
	}

	if x%10 == 0 {
		return false
	}

	temp := x
	reversed := 0

	for temp > 0 {
		reversed = reversed*10 + temp%10
		temp /= 10
		if temp/10 == reversed || temp == reversed {
			return true
		}
	}

	return x == reversed
}
